llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp

```diff
@@ -1075,16 +1003,110 @@ static void simplifyRecipe(VPRecipeBase &R, VPTypeAnalysis &TypeInfo) {
              m_DerivedIV(m_SpecificInt(0), m_SpecificInt(0), m_VPValue()))) &&
       TypeInfo.inferScalarType(R.getOperand(1)) ==
           TypeInfo.inferScalarType(R.getVPSingleValue()))
-    return R.getVPSingleValue()->replaceAllUsesWith(R.getOperand(1));
+    return R.getOperand(1);
+
+  return nullptr;
 }
 
 void VPlanTransforms::simplifyRecipes(VPlan &Plan, Type &CanonicalIVTy) {
   ReversePostOrderTraversal<VPBlockDeepTraversalWrapper<VPBlockBase *>> RPOT(
       Plan.getEntry());
   VPTypeAnalysis TypeInfo(&CanonicalIVTy);
+  SetVector<VPRecipeBase *> Worklist;
+  for (VPBasicBlock *VPBB : VPBlockUtils::blocksOnly<VPBasicBlock>(RPOT))
+    for (VPRecipeBase &R : make_early_inc_range(*VPBB))
+      Worklist.insert(&R);
+
+  while (!Worklist.empty()) {
+    VPRecipeBase *R = Worklist.pop_back_val();
+    if (VPValue *Result = simplifyRecipe(*R, TypeInfo)) {
+      R->getVPSingleValue()->replaceAllUsesWith(Result);
+      R->eraseFromParent();
```

Comment:
I think erasing R might invalidate the VTypeAnalysis cache, I'll take a look to see if this is a problem

Reply:
Not sure if we need eraseFromParent, as there is recursivelyRemoveDeadRecipes.

Reply:
I've manually erased the deleted value from the cache in 4e05ab9c58cae87981a2ad07aaaf322ec5933a8e.

This was an existing issue with the simplification pass but it looks like it didn't trigger in the wild. I think it's probably easiest to fix it in this PR now that we erase all replaced recipes, which should prevent #120252.

I've also moved the VTypeAnalysis assertion up so it should run for all transforms, not just the trunc combine.

Reply:
@artagnon erasing the old instruction decreases the number of uses which can affect what transforms kick in, e.g. the blend simplifications are sensitive to the number of uses.

Also if the old recipe uses the new result, it will get added back to the worklist which will loop

Reply:
I should mention I took the structure from this from InstructionCombining.cpp! There's a few things there that I think we could eventually copy over, e.g. adding any operands of the recipe to the worklist now that they have one less use.

Reply:
I think it would be good if we would have a way to verify we aren't stuck in an infinite cycle.

Not 100% sure how instcombine checks for that, but we could add the new users to simplify in a worklist and switch to it when we are done with the current wordlist and limit the times we switch.